### ISSUE TITLE
Transaction cannot proceed: STATUS_MARKED_ROLLBACK

### ISSUE THREAD

**Issue by PoloShock** [jberet/jsr352#84]
Hi, I am experiencing the above error which is very similar to the following issue: https://issues.jboss.org/browse/JBERET-179

When I am reading or processing an item within chunk, the transaction might be marked as roll back only somewhere in bussiness method due to an exception. When the exception is skippable, JBeret does not take any action in case of readItem() nor processItem() methods in ChunkRunner and continues to use the transaction marked for rollback in another chunk round.

The expected behavior would be, similarly to what is done in doCheckpoint(), to roll back the transaction and to begin it again before the new round of the chunk starts:
```
if (tm.getStatus() == Status.STATUS_MARKED_ROLLBACK) {
	tm.rollback();
	tm.begin();
}
```

**Comment by PoloShock** [comment]
To add some contemplation on the proposed behavior above:
- It would mean potentially having more transactions than one throughout the chunk.
- The intermediate work done before the transaction roll back would be lost.
- Maybe starting a new chunk would be better conceptually? But it would imply not honoring the checkpoint algorithm's isReadyToCheckpoint() logic.
- Or should the "Transaction cannot proceed: STATUS_MARKED_ROLLBACK" exception be managed completely outside of JBeret?

**Comment by chengfang** [comment]
Thanks for the excellent analysis.  If we only rollback the rollback-only transaction either at the beginning (as your suggested above), or at the end (JBeret currently does) of a chunk processing, it should be safe.  For the beginning case, there is no work being done yet; and for the end case, it's the intent of marking it rollback-only to discard any intermediate work.

Starting a new chunk, or handling it outside the framework would deviate from teh spec required behavior.  So I think what you proposed above should work.

**Comment by PoloShock** [comment]
@chengfang I finally got to test the fix on 1.3.1.Final. Unfortunately the problem is still present, since the `processingInfo.chunkState`, which is queried in the enclosing if statement, is at state `RUNNING`, so it never reaches the commited code.

I am testing it by marking a transaction for rollback and rising a skippable exception in `ItemProcessor.processItem` method.

**Comment by chengfang** [comment]
Have you tried the most recent maintenance release 1.3.3.Final?   There are a couple of bug fixes since 1.3.1.Final.

**Comment by PoloShock** [comment]
No I haven't. I am running the code on WildFly 15.0.1.Final which has the 1.3.1.Final jberet version.

**Comment by chengfang** [comment]
jberet-core 1.3.3.Final should be able to drop in and replace your current version in WildFly 15.0.1.  There were a couple of fixes related to chunk processing and transaction handling, and you may want to give it a try.

**Comment by PoloShock** [comment]
I just tried the same setup with 1.3.3.Final and the result is the same as I stated previously (https://github.com/jberet/jsr352/issues/84#issuecomment-459785579).

**Comment by chengfang** [comment]
Thanks for trying this.  I'll take a closer look.

**Comment by chengfang** [comment]
> since the `processingInfo.chunkState`, which is queried in the enclosing `if` statement, is in state `RUNNING`, so it never reaches the commited code.

Can you more specific?  Which line of code in ChunkRunner class?
https://github.com/jberet/jsr352/blob/master/jberet-core/src/main/java/org/jberet/runtime/runner/ChunkRunner.java

**Comment by chengfang** [comment]
Copied this issue to JIRA: [JBERET-471](https://issues.jboss.org/browse/JBERET-471) 
Skip items causing Transaction cannot proceed: STATUS_MARKED_ROLLBACK

**Comment by chengfang** [comment]
Do you have a reproducer test app?  If so can you attach to [JBERET-471](https://issues.jboss.org/browse/JBERET-471) ?

**Comment by chengfang** [comment]
> The expected behavior would be, similarly to what is done in doCheckpoint(), to roll back the transaction and to begin it again before the new round of the chunk starts:

We can apply the similar logic in doCheckpoint() to readItem() and processItem() methods.  That should achieve the same effect as in doCheckpoint().  But be careful with this kind of use case that invovles both skippable exception and rollback-only transaction.  Any valid items that are already processed in the current chunk will be discarded as there is no way to commit the current transaction, and we will not retry them.

**Comment by PoloShock** [comment]
Yes, I am aware of that. My use-case is having a chunk size = 1.

**Comment by PoloShock** [comment]
> > since the `processingInfo.chunkState`, which is queried in the enclosing `if` statement, is in state `RUNNING`, so it never reaches the commited code.
> 
> Can you more specific? Which line of code in ChunkRunner class?
> https://github.com/jberet/jsr352/blob/master/jberet-core/src/main/java/org/jberet/runtime/runner/ChunkRunner.java

I was referring to the commit https://github.com/jberet/jsr352/commit/e0618f51208e0de35ef562ee28bd4501484536c9. In current `master` branch the enclosing `if` is on line number 311.

> Do you have a reproducer test app?

I don't have a reproducer test app, but the situation should be easily reproducible by marking transaction for rollback and rising a skippable exception in `ItemProcessor.processItem` method, as I commented earlier.

**Comment by chengfang** [comment]
the above revision should fix it, passed my tests.  It will be included in the next release jberet-core 1.4.0.Beta1.
